Rewrite of the Stamina skill within osu!taiko

Lawtrohux opened the pull request:
This rewrite of the stamina skill is the first PR regarding the large scale change across the entire osu!taiko difficulty calculation system. Due to the large nature of the rework, it will be released in per skill/ per part.

Some values may seem overweighted, however these are addressed within future pr's as this is the baseline of our changes, and work alongside the future reworks of rhythm, colour, strain and an addition of a new reading skill.

The changes in this PR are outlined within this [general document](
https://docs.google.com/document/d/1Z5GC4DMqOVzeIERMSK3qpQaqjq-sVnhbuoxAwy9qxDs/edit#heading=h.lsi2mua6n53x). 

However, as a basis this rework aims to calculate and apply the theoretical minimum stamina required to play a given pattern with any playstyle, with any theoretically rollable pattern being rolled, rather than assuming a full alternate playstyle. The StaminaCheesePenalty is also removed, as this is being addressed within the colour rework.

Lawtrohux committed:
Stamina Skill Rewrite and Refactor

Lawtrohux commented:
@smoogipoo is it possible to get a smoogisheet for this build?

smoogipoo commented:
SR/PP changes (as of 2022-05-01): https://docs.google.com/spreadsheets/d/1k7i9hNF0O4V8pihktfTmiFAaSCB7SPPfR3EpZ-4jP9I/edit#gid=0

smoogipoo commented:
As requested, spreadsheet without converts: https://docs.google.com/spreadsheets/d/10fxh2VNbNc4uNbYI44a8E8rVFPZhlVm3y1WQ6co3ir4/edit?usp=sharing

Lawtrohux committed:
Remove Redundant Detection

Not needed, as nerf is not required anymore and has negative effect.

Lawtrohux committed:
Refactor and Cleanup of Stamina.cs

Lawtrohux committed:
Include Convert Penalty

For low colour variance

Lawtrohux committed:
Rewrite Performance difficultyValue Calculation

To change values in line with the community survey, pp has been rescaled immensely, with changes on both the lower and upper end.

vunyunt reviewed (dismissed):
On behalf of the taiko pp committee the changes to difficulty and performance are approved.

bdach commented on osu.Game.Rulesets.Taiko/Difficulty/Skills/Stamina.cs:
```diff
@@ -1,16 +1,48 @@
+    /// </summary>
+    public class SingleKeyStamina
+    {
+        private double previousHitTime = -1;
```
Minor code quality note: I would have probably used a `double?` for this rather than a plain `double` and a -1 marker value.

Lawtrohux replied:
Changed in latest commit, removing marker value

bdach commented on osu.Game.Rulesets.Taiko/Difficulty/Skills/Stamina.cs:
```diff
@@ -1,16 +1,48 @@
+                return 0;
+            }
+
+            double objectStrain = 0.5;
+            objectStrain += speedBonus(current.StartTime - previousHitTime);
```
I'm skimming the document linked in the OP, and it says this in the section pointing out issues with the current stamina setup:

> The base rewarding density is also an issue, with a base object strain of 1, density is rewarded at a fixed value rather than being able to calculate density per object at a differential.
The speed bonus is based on the duration of recent objects in the playfield, and thus added to the base object strain.

This doesn't seem to be far off from that, except that the values are slightly different now, which leaves me wondering - is this a temporary state of affairs, or is the doc outdated in some respect? Is this change supposed to improve things with respect to how strains are calculated?

Lawtrohux replied:
I'll update the document to the reflect the current changes now. Strain also is in the reworks of being rewritten, which is what I was referring to in the document and was kept out of this pr.

bdach commented on osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs:
```diff
@@ -69,17 +67,21 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
             staminaRating *= staminaPenalty;
 
-            double combinedRating = locallyCombinedDifficulty(colour, rhythm, staminaRight, staminaLeft, staminaPenalty);
+            if (beatmap.BeatmapInfo.Ruleset.OnlineID == 0 && colourRating < 0.05)
+            {
+                staminaPenalty *= 0.25;
+            }
```
Two questions here:

* Why are converts getting directly targeted like this only in regards to one skill? I believe I've heard things that taiko players don't really value converts, but I'd still like to believe that a map is a map is a map, and it shouldn't matter all that much if it's a convert or not if there's a good enough algorithm for all maps. The commit message mentions "low colour variance" - so is this just some sort of holdover until the colour skill is redone, to ensure that values don't look bad?
* After these lines, `staminaPenalty` is only ever again used once, in `locallyCombinedDifficulty()`, while `separatedRating` uses `staminaRating` which was multiplied by `staminaPenalty` _before_ it had the convert nerf applied to it. Why is that the case?

Lawtrohux replied:
This is a holdover until the colour skill is done, and do fix the issues with monos (single colour) streams. In terms of its placement, it's more of a blanket fix and was utilised in total calculations rather than seperate to be easily controlled. I can fix that up if you'd like however for code quality

bdach replied:
I don't mind the placement if it is indeed temporary, at most you can leave a TODO that this will be addressed later.

Lawtrohux replied:
I'll add a todo, on an other note i'll resolve the code quality primary too as soon as I can.

Lawtrohux replied:
todo added, resolving for now

bdach reviewed (commented):
Code style is overall good, can't see much fault from a correctness standpoint, but I do have some follow-up questions.

Lawtrohux committed:
Apply Review Changes

bdach commented on osu.Game.Rulesets.Taiko/Difficulty/Skills/Stamina.cs:
```diff
@@ -1,16 +1,48 @@
+        /// </summary>
+        public double StrainValueOf(DifficultyHitObject current)
+        {
+            if (previousHitTime == -1)
```
Uh... this condition will never be hit now after the latest commit removed the initialisation of this to -1. What I intended to suggest with my previous review comment is to do this:

```csharp
double? previousHitTime;

// in StrainValueOf():
if (previousHitTime == null)
{
    previousHitTime = current.StartTime;
    return 0;
}
// rest of code follows
```

Lawtrohux replied:
Apologies, was in a bit of a rush and forgot to initiate a `double?`

Lawtrohux committed:
Rectify Review Changes

Lawtrohux commented:
Git autopushed without secondary changes, reverting back now, apologies!

Lawtrohux committed:
Rectify null value call

tired programming is never a good idea

smoogipoo committed:
Adjust tests

smoogipoo commented:
Spreadsheets updated:
**With** converts: https://docs.google.com/spreadsheets/d/1k7i9hNF0O4V8pihktfTmiFAaSCB7SPPfR3EpZ-4jP9I/edit#gid=0
**Without** converts: https://docs.google.com/spreadsheets/d/10fxh2VNbNc4uNbYI44a8E8rVFPZhlVm3y1WQ6co3ir4/edit#gid=0

@ppy/taiko-pp-commitee Please recheck/reapprove.

vunyunt reviewed (dismissed):
The new changes to performance are approved

Lawtrohux commented:
The force push was to rebase to upstream, by accident. No further changes have been made, and are good to go.

bdach reviewed (dismissed):
no objections from me, still the same diff that i read before the accidental rebase/force-push

smoogipoo committed:
Split out SingleKeyStamina class

smoogipoo committed:
Refactor key indexing

smoogipoo commented on osu.Game.Rulesets.Taiko/Difficulty/Skills/SingleKeyStamina.cs:
```diff
@@ -0,0 +1,42 @@
+        /// <param name="notePairDuration">The duration between the current and previous note hit using the same key.</param>
+        private double speedBonus(double notePairDuration)
+        {
+            return 175 / (notePairDuration + 100);
```
I would like to see these constants better defined and documented, probably in terms of a BPM at which the `objectStrain` falls below `1`. This seems to be at exactly 240 BPM (or 250ms between hitobjects) by my calculations.
